src/rgw/rgw_sal_motr.cc

```diff
@@ -895,9 +895,8 @@ int MotrBucket::check_quota(const DoutPrefixProvider *dpp, RGWQuotaInfo& user_qu
 
 int MotrBucket::merge_and_store_attrs(const DoutPrefixProvider *dpp, Attrs& new_attrs, optional_yield y)
 {
-  for (auto& it : new_attrs)
```

Comment:
explain with example how currently this loop is not serving to store updated attributes

Reply:
In breif, Since we entirely removing the tagset key from attrs. Merging new attrs and existing attrs won't work as expected for all bucket delete attribute operations. Because as per the existing logic if a key not present in new passed attribue then the program will consider the existing key value pairs itself.

Reply:
In Detail,
In the existing implementation we are getting available bucket attributes using below code.
rgw::sal::Attrs attrs = s->bucket->get_attrs();

For example consider above new attrs as shown below with tag-attrs key:
{
'acl': <some buffer list>,
'cors': <some buffer list>,
'tag-attrs': <some buffer list> 
} 

After that we are deleting the tag-attrs key as shown below
attrs.erase(RGW_ATTR_TAGS);

so, here is the updated attrs map that will have all available bucket level attributes except tag-attrs since we deleted it.

For example consider new_attrs as below after tag-attrs deleted:
{
'acl': <some buffer list>,
'cors': <some buffer list>
} 

Then the new attrs was passed to merge_and_store_attrs as shown below
op_ret = s->bucket->merge_and_store_attrs(this, attrs, y);

But, In merge_and_store, we again looping the attrs map and trying to merge all the attributes as shown below.

for (auto& it : new_attrs)
	attrs[it.first] = it.second;

Here, new_attrs doesn't hold tag-attrs in it. Since we deleted tag-attrs key using attrs.erase(RGW_ATTR_TAGS). So the loop will run only two times (<0, acl>, <1, cors>) and updating the tag-attrs key will be skipped since new_attrs doesn't have a tag key in it.

Reply:
agree

Reply:
This function name itself indicates that it is merge and store. We should double check if there is any case where we are partially initializing new_attrs and merging it into the existing attrs.

Reply:
Hi @sjain09, Got your point. But we have a validation scenario here for tags saying **NoSuchTagSetError** which will be raised where there is no tags attached to that bucket.

Say, we have one tag here
```
[root@ssc-vm-g3-rhev4-2997 build]# aws --endpoint=$endpoint s3api get-bucket-tagging --bucket test-obj-bucket
{
    "TagSet": [
        {
            "Key": "@#aws:designation",
            "Value": "confidential"
        }
    ]
}
```
Then we deleting it and again trying to get the tags then below validation message should be raised
```
[root@ssc-vm-g3-rhev4-2997 build]# aws --endpoint=$endpoint s3api delete-bucket-tagging --bucket test-obj-bucket
[root@ssc-vm-g3-rhev4-2997 build]# aws --endpoint=$endpoint s3api get-bucket-tagging --bucket test-obj-bucket

An error occurred (**NoSuchTagSetError**) when calling the GetBucketTagging operation: Unknown

```
above error message will be raised only when the entire tagset key, value is removed from bucket attributes.

So, for merging some two objects K,V we need same key should be present in both objects which need to be merged. which is not possible in our case. because we have to remove that entire key to make existing validation work.

Reply:
on other hand trying to handle this work around in merging way itself will affect the generic method.

Note: Actually i have tested few more other API 's to ensure this fix. Say when ever you are deleting some bucket/ object/ cors/ other related attributes and calling merge_and_save it won't work because of existing code. After this fix i can able to ensure that those are working fine as expected.

@sjain09